// @nuxtjs/eslint-module/dist/module.mjs
import { defineNuxtModule, addVitePlugin, addWebpackPlugin } from '@nuxt/kit';
import vitePluginEslint from 'vite-plugin-eslint';
import EslintWebpackPlugin from 'eslint-webpack-plugin';

const name = "@nuxtjs/eslint-module";
const version = "4.0.2";

const module = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "eslint",
    compatibility: {
      bridge: true
    }
  },
  defaults: (nuxt) => ({
    cache: true,
    include: [`${nuxt.options.srcDir}/**/*.{js,jsx,ts,tsx,vue}`],
    exclude: ["**/node_modules/**", nuxt.options.buildDir],
    eslintPath: "eslint",
    formatter: "stylish",
    lintOnStart: true,
    emitWarning: true,
    emitError: true,
    failOnWarning: false,
    failOnError: false
  }),
  setup(options, nuxt) {
    if (!nuxt.options.dev) {
      return;
    }
    addVitePlugin(vitePluginEslint(options), { server: false });
    const webpackOptions = {
      ...options,
      context: nuxt.options.srcDir,
      files: options.include,
      lintDirtyModulesOnly: !options.lintOnStart
    };
    delete webpackOptions.include;
    delete webpackOptions.lintOnStart;
    addWebpackPlugin(new EslintWebpackPlugin(webpackOptions), { server: false });
  }
});

export { module as default };
